aesd-char-driver/main.c

```diff
@@ -0,0 +1,278 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h>   // file_operations
+#include <linux/slab.h>
+#include <linux/string.h>
+#include "aesdchar.h"
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Shashank Chandrasekaran"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+    PDEBUG("open");
+    /**
+     * TODO: handle open
+     */
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                  loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    int tmp_buffer_count = 0;
+    size_t offset_byte;
+    struct aesd_buffer_entry *tmp_buffer;
+    struct aesd_dev *dev;
+
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    /**
+     * TODO: handle read
+     */
+
+    dev = (struct aesd_dev*)filp->private_data;
+
+    if (mutex_lock_interruptible(&dev->lock)!=0)
+	{
+		PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+		goto handle_error;
+	}
+
+    tmp_buffer = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->circular_buffer, *f_pos, &offset_byte);
+
+    if(tmp_buffer==NULL)
+        goto handle_error;
+
+    if ((tmp_buffer->size - offset_byte) < count) 
+    {
+        *f_pos = *f_pos + (tmp_buffer->size - offset_byte);
```

Comment:
@shashankchandrasekaran you might want to adjust the file position after copy_to_user so you don't adjust f_pos assuming success and then have an error condition.